control-plane-operator/controllers/hostedcontrolplane/nodebootstrapper/kubeconfig.go

```diff
@@ -0,0 +1,195 @@
+package nodebootstrapper
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/openshift/hypershift/control-plane-operator/controllers/hostedcontrolplane/manifests"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	clientcmdv1 "k8s.io/client-go/tools/clientcmd/api/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/yaml"
+
+	hyperv1 "github.com/openshift/hypershift/api/v1alpha1"
+)
+
+const (
+	KubeconfigKey                  = "kubeconfig"
+	machineConfigOperatorNamespace = "openshift-machine-config-operator"
+)
+
+// ReconcileBootstrapKubeconfig maintains a secret named bootstrap-kubeconfig. This secret will be used in kubelet client CSR creation.
+// The fucntion uses a kubeconfig in the service-network-admin-kubeconfig secret to connect to the hosted cluster.
+// On the hosted cluster, a node-bootstrapper service account in the openshift-machine-config-operator namespace is created.
+// The service account token will be used to generate a kubeconfig named bootstrap-kubeconfig on the management cluster, which kubelet will use for client csr.
+func ReconcileBootstrapKubeconfig(mangementClient client.Client, ctx context.Context, hcp *hyperv1.HostedControlPlane) error {
+	if hcp.Status.KubeConfig == nil {
+		return fmt.Errorf("hosted controlplane kubeconfig secret is not generated")
+	}
+	// Resolve the kubeconfig secret for the hosted control plane
+	kubeconfigSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: hcp.Namespace,
+			Name:      "service-network-admin-kubeconfig",
+		},
+	}
+	err := mangementClient.Get(ctx, client.ObjectKeyFromObject(kubeconfigSecret), kubeconfigSecret)
+	if err != nil {
+		return fmt.Errorf("failed to get hosted controlplane kubeconfig secret %q: %w", kubeconfigSecret.Name, err)
+	}
+
+	// construct client for hosted controlplane
+	var config *clientcmdapi.Config
+
+	if kubeconfig, ok := kubeconfigSecret.Data[KubeconfigKey]; ok {
+		config, err = clientcmd.Load(kubeconfig)
+		if err != nil {
+			return err
+		}
+	}
+	if config == nil {
+		return fmt.Errorf("failed to load hosted controlplane kubeconfig from secret %q: %w", kubeconfigSecret.Name, err)
+	}
+	clientConfig, err := clientcmd.NewDefaultClientConfig(*config, &clientcmd.ConfigOverrides{}).ClientConfig()
+	if err != nil {
+		return fmt.Errorf("failed to create hosted controlplane client from secret %q: %w", kubeconfigSecret.Name, err)
+	}
+	hostedClusterClient, err := client.New(clientConfig, client.Options{})
```

Comment:
why do we need to do anything in the guest cluster? Isn't all we need to adjust https://github.com/openshift/hypershift/blob/a86c74d2528991839e09725660b56dfa5fddbca3/control-plane-operator/controllers/hostedcontrolplane/pki/kas.go#L52?